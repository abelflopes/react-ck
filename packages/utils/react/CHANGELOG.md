# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.6.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.5.0...@react-ck/react-utils@1.6.0) (2025-07-09)


### Features

* add expandable component ([e5cd8e7](https://github.com/abelflopes/react-ck/commit/e5cd8e7c3362eb80bf798ffa83757b35d9b43df2))



## [1.5.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.4.0...@react-ck/react-utils@1.5.0) (2025-06-07)


### Features

* support focus trapping and apply consistent focus outline ([26ea216](https://github.com/abelflopes/react-ck/commit/26ea2167102a4d1c89f9bd937dc44949ef2a8b0e))



## [1.4.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.7...@react-ck/react-utils@1.4.0) (2024-12-19)


### Features

* **react utils:** add ref merge util ([42683b3](https://github.com/abelflopes/react-ck/commit/42683b3cf99617076eb29588fd6b6d94bf842b2f))



## [1.3.7](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.6...@react-ck/react-utils@1.3.7) (2024-10-09)


### Bug Fixes

* update react version ([59e4e9a](https://github.com/abelflopes/react-ck/commit/59e4e9afa979d29efdc793f3441ed528971844ca))



## [1.3.6](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.5...@react-ck/react-utils@1.3.6) (2024-07-10)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.5](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.4...@react-ck/react-utils@1.3.5) (2024-07-10)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.4](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.3...@react-ck/react-utils@1.3.4) (2024-07-09)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.3](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.2...@react-ck/react-utils@1.3.3) (2024-07-08)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.2](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.1...@react-ck/react-utils@1.3.2) (2024-07-05)

**Note:** Version bump only for package @react-ck/react-utils





## [1.3.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.3.0...@react-ck/react-utils@1.3.1) (2024-07-05)


### Bug Fixes

* component identification with display name ([b182afb](https://github.com/abelflopes/react-ck/commit/b182afb3468d0ca00d9923f05d678bc8db8492f0))



## [1.3.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.12...@react-ck/react-utils@1.3.0) (2024-07-04)


### Features

* add click outside util ([738a2aa](https://github.com/abelflopes/react-ck/commit/738a2aab13db43fbe4a64f41030ec40fd6fa7074))
* add next render util ([30ca88a](https://github.com/abelflopes/react-ck/commit/30ca88a03ec1c760097331a4f48857b0d7b017d2))
* update react utils ([75c4350](https://github.com/abelflopes/react-ck/commit/75c43505ddba5382cc71e619588764abf179d84f))



## [1.2.12](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.11...@react-ck/react-utils@1.2.12) (2024-06-26)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.11](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.10...@react-ck/react-utils@1.2.11) (2024-06-23)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.10](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.9...@react-ck/react-utils@1.2.10) (2024-06-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.9](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.8...@react-ck/react-utils@1.2.9) (2024-03-19)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.8](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.7...@react-ck/react-utils@1.2.8) (2024-03-17)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.7](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.6...@react-ck/react-utils@1.2.7) (2024-03-17)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.6](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.5...@react-ck/react-utils@1.2.6) (2024-03-16)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.5](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.4...@react-ck/react-utils@1.2.5) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.4](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.3...@react-ck/react-utils@1.2.4) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.3](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.2...@react-ck/react-utils@1.2.3) (2024-03-15)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.2](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.1...@react-ck/react-utils@1.2.2) (2024-03-12)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.2.0...@react-ck/react-utils@1.2.1) (2024-03-12)

**Note:** Version bump only for package @react-ck/react-utils





## [1.2.0](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.1.1...@react-ck/react-utils@1.2.0) (2024-03-06)


### Features

* add component polimorphy util ([e653799](https://github.com/abelflopes/react-ck/commit/e65379984d022ca716a3b040aae03aa65528fcaa))
* add util to convert component into text ([4580f6c](https://github.com/abelflopes/react-ck/commit/4580f6c4630d179b2dda084cfaa04c518f551efc))


### Bug Fixes

* child list misssing key warning ([e05e301](https://github.com/abelflopes/react-ck/commit/e05e3019493b01bf492aa13de775adc7891faa35))



## [1.1.1](https://github.com/abelflopes/react-ck/compare/@react-ck/react-utils@1.1.0...@react-ck/react-utils@1.1.1) (2024-01-24)

**Note:** Version bump only for package @react-ck/react-utils





## 1.1.0 (2024-01-02)


### Features

* add react utils package ([fe8737f](https://github.com/abelflopes/react-ck/commit/fe8737f2bf784a0b42111a03980112fa3d4f0795))


### Bug Fixes

* make packages public ([16c94d3](https://github.com/abelflopes/react-ck/commit/16c94d3bd288a46084896f7fa6f2be33a3609e4f))
