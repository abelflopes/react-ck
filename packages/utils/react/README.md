# React CK | Utils Library

> :warning: **WARNING**: This component library is being updated frequently and it's currently unstable due to being in it's early stages, we advice you to use only in production environments only after version **2.0.0**.

### Installation 

To integrate the this library into your React apps, you can install it using npm or yarn: `npm i --save @react-ck/react-utils` or `yarn add @react-ck/react-utils`.

## Utilities

### Polimorphy

Component polimorphy allows consumer to choose which HTML tags and attributes will be rendered from a specific component.

### Usage Examples

### Implementation

```diff
import React, { isValidElement, useEffect, useMemo } from "react";
import styles from "./styles/index.module.scss";
import classNames from "classnames";
- import { getDisplayName } from "@react-ck/react-utils";
+ import {
+  getDisplayName,
+  PolymorphicComponent,
+  type ConsumerPolymorphicProps,
+  type HTMLTag,
+ } from "@react-ck/react-utils";

- export interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+ export interface ButtonProps<T extends HTMLTag>
+  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
+    ConsumerPolymorphicProps<T> {
  skin?: "primary" | "secondary" | "ghost" | "negative";
  size?: "s" | "m" | "l";
  icon?: React.ReactNode;
}

- export const Button = ({
+ export const Button = <T extends HTMLTag>({
+  as,
  skin = "primary",
  size = "m",
  icon,
  children,
  className,
  ...otherProps
- }: Readonly<ButtonProps>): React.ReactElement => {
+ }: Readonly<ButtonProps<T>>): React.ReactElement => {
  const isIconOnly = useMemo(
    () => icon !== undefined && React.Children.count(children) === 0,
    [children, icon],
  );

  return (
-    <button
-      className={classNames(
-        styles.root,
-        styles[skin],
-        styles[`size-${size}`],
-        {
-          [`${styles["icon-only"]}`]: isIconOnly,
-        },
-        className,
-      )}
-      {...otherProps}>
+    <PolymorphicComponent
+      as={as}
+      fallback={["button", otherProps]}
+      commonProps={{
+        className: classNames(
+          styles.root,
+          styles[skin],
+          styles[`size-${size}`],
+          {
+            [`${styles["icon-only"]}`]: isIconOnly,
+          },
+          className,
+        ),
+      }}>
      {icon && !isIconOnly ? <span className={styles.icon}>{icon}</span> : null}
      {children}
      {isIconOnly ? icon : null}
-     </button>
+    </PolymorphicComponent>
  );
};
```


<!-- storybook-ignore -->

---

Check the documentation website - [react-ck.js.org](https://react-ck.js.org).