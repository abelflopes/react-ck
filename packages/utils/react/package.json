{
  "name": "@react-ck/react-utils",
  "private": false,
  "version": "1.3.6",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "/dist"
  ],
  "homepage": "https://github.com/abelflopes/react-ck/tree/master/packages/utils/react-hooks#readme",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/abelflopes/react-ck.git"
  },
  "scripts": {
    "build": "tsc -p tsconfig.build.json",
    "lint:typescript": "tsc --noEmit",
    "lint:code": "eslint . --ext ts,tsx --cache",
    "test": "jest --testPathPattern=\".unit.*\""
  },
  "devDependencies": {
    "@react-ck/babel-config": "^1.0.0",
    "@react-ck/jest-config": "^1.0.0",
    "@react-ck/typescript-config": "^1.0.0",
    "@types/react": "^18.3.11"
  },
  "peerDependencies": {
    "react": "^18.3.1"
  },
  "dependencies": {
    "classnames": "^2.5.1"
  }
}
