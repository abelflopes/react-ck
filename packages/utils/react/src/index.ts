export * from "./display-name";

export * from "./polimorphy";

export * from "./component-to-text";

export * from "./click-outside";

export * from "./children-without-fragments";

export * from "./merge-refs";

export * from "./raf";

export * from "./promise";
