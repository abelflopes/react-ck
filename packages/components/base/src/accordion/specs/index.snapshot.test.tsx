import React from "react";
import { Accordion } from "../index";
import { render } from "@testing-library/react";

describe("snapshot accordion", () => {
  it("renders correctly", () => {
    const tree = render(
      <Accordion
        items={[
          { header: "item 1", children: "content 1" },
          { header: "item 2", children: "content 2" },
        ]}
        keepInDom
      />,
    ).asFragment();
    expect(tree).toMatchSnapshot();
  });
});
