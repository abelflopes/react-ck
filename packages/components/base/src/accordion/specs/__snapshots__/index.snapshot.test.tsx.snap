// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`snapshot accordion renders correctly 1`] = `
<DocumentFragment>
  <ul
    class="root bordered"
  >
    <li
      class="item"
    >
      <details
        class="spacing_s"
      >
        <summary
          class="header"
        >
          <span
            class="VsR_m2r2hGIr7NfX85BU YEVFelv_bfcF2Ee4WG8g sf93RNkSSnS5VkGCtSLh"
          >
            <svg
              class="icon"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.34317 7.75732L4.92896 9.17154L12 16.2426L19.0711 9.17157L17.6569 7.75735L12 13.4142L6.34317 7.75732Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="header_content"
          >
            item 1
          </div>
        </summary>
        <div
          class="content"
        >
          content 1
        </div>
      </details>
    </li>
    <li
      class="item"
    >
      <details
        class="spacing_s"
      >
        <summary
          class="header"
        >
          <span
            class="VsR_m2r2hGIr7NfX85BU YEVFelv_bfcF2Ee4WG8g sf93RNkSSnS5VkGCtSLh"
          >
            <svg
              class="icon"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6.34317 7.75732L4.92896 9.17154L12 16.2426L19.0711 9.17157L17.6569 7.75735L12 13.4142L6.34317 7.75732Z"
                fill="currentColor"
              />
            </svg>
          </span>
          <div
            class="header_content"
          >
            item 2
          </div>
        </summary>
        <div
          class="content"
        >
          content 2
        </div>
      </details>
    </li>
  </ul>
</DocumentFragment>
`;
